File: frontend/src/lib/state.ts
import { invalidateButDeferUntilNavigation } from './notification';
import { fetchWithNotify } from './fetchWithNotify';

export type ModuleSettings = {
	type: string;
	settings: {
		[key: string]: unknown;
	};
	/** Per-setting priority overrides, keyed by setting name (lower wins). */
	priorities?: {
		[key: string]: number;
	};
};

/**
 * Priority used for module settings defined directly on a configuration (device).
 * Mirrors HOST_PRIORITY in controller/thymis_controller/lib.py.
 */
export const HOST_PRIORITY = 80;

export type Origin = {
	originId: string;
	originContext: string;
	originName: string;
};

export type ModuleSettingsWithOrigin = {
	type: string;
	settings: {
		[key: string]: unknown;
	};
	/** Per-setting priority overrides of this definition (lower wins). */
	priorities?: {
		[key: string]: number;
	};
	originId: string;
	originContext: string;
	originName: string;
	priority: number | undefined;
};

export type SecretType = 'single_line' | 'multi_line' | 'env_list' | 'file';
export type SecretProcessingType = 'none' | 'mkpasswd-yescrypt';

export type Secret = {
	id: string;
	display_name: string;
	type: SecretType;
	value_str: string | null;
	value_size: number;
	filename: string | null;
	include_in_image: boolean;
	processing_type: SecretProcessingType;
	error?: string | null;
	created_at: string;
	updated_at: string;
	delete_at: string | null;
};

export type SelectOneSettingType = {
	'select-one': [string, string][];
	'extra-data'?: Record<string, unknown>;
};

export type ListSettingType = {
	'list-of': Record<string, Setting>;
	'element-name'?: string;
};

export type SecretSettingType = {
	type: 'secret';
	'allowed-types'?: SecretType[];
	'default-processing-type'?: SecretProcessingType;
	'default-save-to-image'?: boolean;
};

export type ArtifactSettingType = {
	type: 'artifact';
};

export type TextAreaCodeSettingType = {
	type: 'textarea-code';
	language?: string;
};

export type SystemdTimerSetting = {
	timer_type?: 'realtime' | 'monotonic' | 'continuous';
	on_boot_sec?: string;
	on_unit_active_sec?: string;
	accuracy_sec?: string;
	on_calendar?: string[];
	persistent?: boolean;
	randomized_delay_sec?: string;
};

export type SystemdTimerSettingType = {
	type: 'systemd-timer';
} & SystemdTimerSetting;

export type SettingType =
	| 'string'
	| 'number'
	| 'bool'
	| 'textarea'
	| 'int'
	| SelectOneSettingType
	| ListSettingType
	| SecretSettingType
	| ArtifactSettingType
	| TextAreaCodeSettingType;

export type Setting<T extends SettingType = SettingType> = {
	displayName: string;
	type: T;
	description?: string | null;
	default?: unknown | null;
	example?: string | null;
	order?: number;
	/** Whether this setting's nix definitions honour a per-setting priority override. */
	priorityOverridable?: boolean;
};

export type Module = {
	type: string;
	icon: string | undefined;
	iconDark: string | undefined;
	displayName: string;
	description?: string | null;
	category?: string | null;
	settings: Record<string, Setting>;
};

export type Tag = {
	displayName: string;
	identifier: string;
	priority: number;
	modules: ModuleSettings[];
};

export type Config = {
	displayName: string;
	identifier: string;
	modules: ModuleSettings[];
	tags: string[];
};

export type Repo = {
	url: string;
	api_key_secret: string | null;
};

export type State = {
	version: string;
	repositories: { [name: string]: Repo };
	configs: Config[];
	tags: Tag[];
};

export const saveState = async (state: State) => {
	const response = await fetchWithNotify(`/api/state`, {
		method: 'PATCH',
		headers: {
			'content-type': 'application/json'
		},
		body: JSON.stringify({
			version: state.version,
			repositories: state.repositories,
			configs: state.configs,
			tags: state.tags
		})
	});
	await invalidateButDeferUntilNavigation((url) => url.pathname === '/api/state');
	return response.ok;
};

export const build = async () => {
	await fetchWithNotify(`/api/action/build`, { method: 'POST' });
};

export const getTagByIdentifier = (state: State, identifier: string | null) => {
	if (!identifier) return undefined;
	return state.tags.find((tag) => tag.identifier === identifier);
};

export const getConfigByIdentifier = (state: State, identifier: string | null) => {
	if (!identifier) return undefined;
	return state.configs.find((config) => config.identifier === identifier);
};

export type ContextType = 'tag' | 'config';

File: controller/thymis_controller/models/state.py
import uuid
from typing import Dict, List, Optional

from pydantic import BaseModel, JsonValue
from thymis_controller import migration


class Repo(BaseModel):
    url: Optional[str] = None
    follows: Optional[str] = None
    inputs_follows: Dict[str, str] = {}
    api_key_secret: Optional[uuid.UUID] = None


class ModuleSettings(BaseModel):
    type: str  # type of module this settings object is for
    settings: Dict[str, JsonValue]
    # Per-setting priority overrides, keyed by setting name. The value is used
    # as the lib.mkOverride priority for the nix definitions emitted for that
    # setting (lowest number wins in the NixOS module system). Settings without
    # an entry use the priority of the tag/config this object belongs to.
    priorities: Dict[str, int] = {}

    def setting_priority(self, setting_key: str, default: int) -> int:
        """Priority to emit this setting with, defaulting to the owning
        tag/config priority when the setting has no override."""
        return self.priorities.get(setting_key, default)


THYMIS_DEVICE_MODULE_TYPE = "thymis_controller.modules.thymis.ThymisDevice"


class Config(BaseModel):
    displayName: str
    identifier: str
    modules: List[ModuleSettings]
    tags: List[str]

    def device_type(self):
        for module in self.modules:
            if module.type == THYMIS_DEVICE_MODULE_TYPE:
                return module.settings.get("device_type")
        return None


class Tag(BaseModel):
    displayName: str
    identifier: str
    priority: int
    modules: List[ModuleSettings]


class State(BaseModel):
    version: str = migration.latest_version
    repositories: Dict[str, Repo] = {}
    tags: List[Tag] = []
    configs: List[Config] = []


__all__ = ["Repo", "ModuleSettings", "Config", "Tag", "State"]
